[V1] [Doc] Update V1 docs for Mamba models

tdoublep opened the pull request:
## Purpose

~This PR automates the choice of attention block size when running hybrid (attention/SSM) models in V1. This simplifies things and unburdens the user from having to select the right block size.~

~This change also enables running the models in V1 via the `vllm serve` CLI.~

This PR just updates the V1 docs to reflect the current status of Mamba support on main. I will create another PR with the automated block size changes. 

## Test Plan

~I've removed fixing of the block size from the existing V1 hybrid tests.~

## Test Result

~Should pass but let's see the CI.~

## (Optional) Documentation Update

**I've updated both the supported models and V1 guide documentation.** 

cc @heheda12345 @DarkLight1337 @tlrmchlsmth

tdoublep committed:
Automate choice of attention block size; update docs

Signed-off-by: Thomas Parnell <tpa@zurich.ibm.com>

tdoublep committed:
Fix typo

Signed-off-by: Thomas Parnell <tpa@zurich.ibm.com>

tdoublep committed:
More verbose logging / asserts

Signed-off-by: Thomas Parnell <tpa@zurich.ibm.com>

tdoublep commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -2613,10 +2613,105 @@ def get_kv_cache_spec(self) -> dict[str, KVCacheSpec]:
+            )
+
+        assert new_attn_page_size == new_mamba_page_size
+
```
Done

tdoublep commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -2613,10 +2613,105 @@ def get_kv_cache_spec(self) -> dict[str, KVCacheSpec]:
+            logger.info(
+                "Setting attention block size to %d tokens "
+                "to ensure that attention page size is >= mamba page size.",
+                attn_block_size)
```
Done

tdoublep committed:
Update docs re: FlashInfer

Signed-off-by: Thomas Parnell <tpa@zurich.ibm.com>

heheda12345 commented on docs/models/supported_models.md:
```diff
@@ -392,7 +392,7 @@ Specified using `--task generate`.
 | `MiniMaxM1ForCausalLM`                        | MiniMax-Text                                        | `MiniMaxAI/MiniMax-M1-40k`, `MiniMaxAI/MiniMax-M1-80k`etc.                                                                                                                       |                        |                             |                       |
 | `MiniMaxText01ForCausalLM`                        | MiniMax-Text                                        | `MiniMaxAI/MiniMax-Text-01`, etc.                                                                                                                                            |                        |                             |                       |
-| `Zamba2ForCausalLM`                               | Zamba2                                              | `Zyphra/Zamba2-7B-instruct`, `Zyphra/Zamba2-2.7B-instruct`, `Zyphra/Zamba2-1.2B-instruct`, etc.                                                                              |                        |                             |                       |
+| `Zamba2ForCausalLM`                               | Zamba2                                              | `Zyphra/Zamba2-7B-instruct`, `Zyphra/Zamba2-2.7B-instruct`, `Zyphra/Zamba2-1.2B-instruct`, etc.                                                                              |                        |                             |✅︎                      |
```
Can you also help to update mamba2?

tdoublep replied:
will do

heheda12345 commented on docs/usage/v1_guide.md:
```diff
@@ -83,7 +83,8 @@ based on assigned priority, with FCFS as a tie-breaker), configurable via the
 | **Encoder-Decoder Models**  | <nobr>🟠 Delayed</nobr>                                                            |
 | **Embedding Models**        | <nobr>🟢 Functional</nobr>                                                         |
-| **Mamba Models**            | <nobr>🚧 WIP ([PR #19327](https://github.com/vllm-project/vllm/pull/19327))</nobr> |
+| **Mamba Models**            | <nobr>🟢 Functional</nobr>                                                         |
+| **Hybrid Models**           | <nobr>🟢 Functional</nobr>                                                         |
```
Still prefer to mark it as WIP as not all mamba models are supported. And prefer not to add a new "hybrid model" entry.

heheda12345 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -2613,10 +2613,107 @@ def get_kv_cache_spec(self) -> dict[str, KVCacheSpec]:
         """
 
-        block_size = self.vllm_config.cache_config.block_size
+        attn_block_size = self.vllm_config.cache_config.block_size
```
Can you keep the code path for non-mamba model simple to help people that don't work on these models?
For example
```python
kv_cache_spec = self._get_kv_cache_spec()
if has_mamba and has_attn:
    self.update_kv_cache_spec_for_hybrid_attention_mamba()
```

heheda12345 reviewed (commented):
`cache_config.block_size` is accessed here and there in vLLM, so we need to find a way to update this values safely. https://github.com/vllm-project/vllm/issues/19407 faces a similar issue about updating `max_model_len`. Maybe you can talk with her.

I don't have  a good idea yet but happy to discuss about it.

tdoublep commented on docs/usage/v1_guide.md:
```diff
@@ -104,8 +105,17 @@ to enable simultaneous generation and embedding using the same engine instance i
+(e.g., `MambaForCausalLM`, `JambaForCausalLM`) are not yet suported. Please note that these models currently require
+enforcing eager mode and disabling prefix caching in V1.
+
+#### Hybrid Models
```
will do

tdoublep commented:
OK! Sounds good, let me give it another shot and get back to you.

tdoublep committed:
Resolve conflicts with main; Update docs based on review feedback

Signed-off-by: Thomas Parnell <tpa@zurich.ibm.com>

DarkLight1337 commented on docs/usage/v1_guide.md:
```diff
@@ -83,7 +83,7 @@ based on assigned priority, with FCFS as a tie-breaker), configurable via the
 | **Encoder-Decoder Models**  | <nobr>🟠 Delayed</nobr>                                                            |
 | **Embedding Models**        | <nobr>🟢 Functional</nobr>                                                         |
-| **Mamba Models**            | <nobr>🚧 WIP (<gh-pr:19327>)</nobr> |
+| **Mamba Models**            | <nobr>🚧 WIP (<gh-pr:19327>, <gh-pr:20016>)</nobr> |
```
We can mark Mamba2 as supported (using the same format as CPU support above)

DarkLight1337 replied:
Yes

tdoublep committed:
Address review feedback

Signed-off-by: Thomas Parnell <tpa@zurich.ibm.com>

tdoublep commented:
@DarkLight1337 @heheda12345 Should we merge this PR to get the doc updates in while I continue working on the config stuff? I don't really think it makes sense to tie it together. 

If you agree, I would revert the non-doc changes and also add a note that hybrid models don't work via CLI yet.

tdoublep committed:
Second attempt at auto-setting attention block size

Signed-off-by: Thomas Parnell <tpa@zurich.ibm.com>

tdoublep commented:
@heheda12345 I took another go the config update. Now everything happens from within `verify_and_update_config`. There are now no changes to the GPU model runner code at all. 

Note I can clean up the if/elif/else stuff when parsing the mamba config easily, but want to get your thoughts on this approach.

tdoublep committed:
Fix logic slightly

Signed-off-by: Thomas Parnell <tpa@zurich.ibm.com>

tdoublep committed:
Cleanup

Signed-off-by: Thomas Parnell <tpa@zurich.ibm.com>

tdoublep committed:
Merge branch 'main' into auto-block-size

tdoublep committed:
Remove unused import

Signed-off-by: Thomas Parnell <tpa@zurich.ibm.com>

heheda12345 commented on vllm/model_executor/models/config.py:
```diff
@@ -191,10 +198,197 @@ def verify_and_update_config(vllm_config: "VllmConfig") -> None:
+    "BambaForCausalLM": HybridAttentionMambaModelConfig,
+    "GraniteMoeHybridForCausalLM": HybridAttentionMambaModelConfig,
+    "NemotronHForCausalLM": NemotronHModelConfig,
+    "Zamba2ForCausalLM": Zamba2ModelConfig,
```
I don't want to require each new model to update this page.  What about letting get_mamba_cache_shape to be an interface that should be implemented by each hybrid model? The abstractions you made now can be some useful utility function to minimize code duplication when implementing this interface for each model.
```python
class IsHybrid(Protocol):
     def get_mamba_cache_shape(cls, ...)
```

tdoublep replied:
Yeah, was thinking of something similar. I will have a go at it and open a new PR.

heheda12345 commented on vllm/model_executor/models/config.py:
```diff
@@ -191,10 +198,197 @@ def verify_and_update_config(vllm_config: "VllmConfig") -> None:
+            use_mla=model_config.use_mla).page_size_bytes
+
+        # get mamba page size
+        mamba_page_size = MambaSpec(
```
As the logic of getting MambaSpec here is different from that in gpu_model_runner, can you add some check in gpu_model_runner.get_kv_cache_spec to verify that these two are the same. 
BTW I think we don't need to check FullAttentionSpec as exceptions will be raised if page size doesn't match.

tdoublep replied:
sure

heheda12345 commented:
> Should we merge this PR to get the doc updates in while I continue working on the config stuff?
I'm OK with that, but please remember to update the doc in your new PR.

tdoublep committed:
Revert other changs; update docs

Signed-off-by: Thomas Parnell <tpa@zurich.ibm.com>

tdoublep commented:
@DarkLight1337 @heheda12345 I've reverted the other changes and added a note about the CLI limitation for the hybrid mamba/attention models. 

I will continue refactoring the auto block size selection and open a new PR.

DarkLight1337 commented on docs/models/supported_models.md:
```diff
@@ -316,7 +316,7 @@ Specified using `--task generate`.
 | `ArcticForCausalLM` | Arctic | `Snowflake/snowflake-arctic-base`, `Snowflake/snowflake-arctic-instruct`, etc. | | ✅︎ | ✅︎ |
 | `BaiChuanForCausalLM` | Baichuan2, Baichuan | `baichuan-inc/Baichuan2-13B-Chat`, `baichuan-inc/Baichuan-7B`, etc. | ✅︎ | ✅︎ | ✅︎ |
-| `BambaForCausalLM` | Bamba | `ibm-ai-platform/Bamba-9B-fp8`, `ibm-ai-platform/Bamba-9B` | ✅︎ | ✅︎ | |
+| `BambaForCausalLM` | Bamba | `ibm-ai-platform/Bamba-9B-fp8`, `ibm-ai-platform/Bamba-9B` | ✅︎ | ✅︎ | ✅ |
```
Can you make the checkmark icon consistent with the other entries in this table?

tdoublep replied:
The checkmark icon looks identical to me. Could you paste a screenshot of the difference you are seeing?

DarkLight1337 replied:
![image](https://github.com/user-attachments/assets/6fbd1177-3591-4581-900a-f815aecd96e4)

DarkLight1337 replied:
It's also visible in the git diff

DarkLight1337 commented on docs/models/supported_models.md:
```diff
@@ -316,7 +316,7 @@ Specified using `--task generate`.
 | `ArcticForCausalLM` | Arctic | `Snowflake/snowflake-arctic-base`, `Snowflake/snowflake-arctic-instruct`, etc. | | ✅︎ | ✅︎ |
 | `BaiChuanForCausalLM` | Baichuan2, Baichuan | `baichuan-inc/Baichuan2-13B-Chat`, `baichuan-inc/Baichuan-7B`, etc. | ✅︎ | ✅︎ | ✅︎ |
-| `BambaForCausalLM` | Bamba | `ibm-ai-platform/Bamba-9B-fp8`, `ibm-ai-platform/Bamba-9B` | ✅︎ | ✅︎ | |
+| `BambaForCausalLM` | Bamba | `ibm-ai-platform/Bamba-9B-fp8`, `ibm-ai-platform/Bamba-9B` | ✅︎ | ✅︎ | ✅ |
```
```suggestion
| `BambaForCausalLM` | Bamba | `ibm-ai-platform/Bamba-9B-fp8`, `ibm-ai-platform/Bamba-9B` | ✅︎ | ✅︎ | ✅︎ |
```

DarkLight1337 commented on docs/models/supported_models.md:
```diff
@@ -357,14 +357,14 @@ Specified using `--task generate`.
 | `LlamaForCausalLM` | Llama 3.1, Llama 3, Llama 2, LLaMA, Yi | `meta-llama/Meta-Llama-3.1-405B-Instruct`, `meta-llama/Meta-Llama-3.1-70B`, `meta-llama/Meta-Llama-3-70B-Instruct`, `meta-llama/Llama-2-70b-hf`, `01-ai/Yi-34B`, etc. | ✅︎ | ✅︎ | ✅︎ |
 | `MambaForCausalLM` | Mamba | `state-spaces/mamba-130m-hf`, `state-spaces/mamba-790m-hf`, `state-spaces/mamba-2.8b-hf`, etc. | | ✅︎ | |
-| `Mamba2ForCausalLM` | Mamba2 | `mistralai/Mamba-Codestral-7B-v0.1`, etc. | | ✅︎ | |
+| `Mamba2ForCausalLM` | Mamba2 | `mistralai/Mamba-Codestral-7B-v0.1`, etc. | | ✅︎ | ✅ |
```
```suggestion
| `Mamba2ForCausalLM` | Mamba2 | `mistralai/Mamba-Codestral-7B-v0.1`, etc. | | ✅︎ | ✅︎ |
```

DarkLight1337 commented on docs/models/supported_models.md:
```diff
@@ -357,14 +357,14 @@ Specified using `--task generate`.
 | `MPTForCausalLM` | MPT, MPT-Instruct, MPT-Chat, MPT-StoryWriter | `mosaicml/mpt-7b`, `mosaicml/mpt-7b-storywriter`, `mosaicml/mpt-30b`, etc. | | ✅︎ | ✅︎ |
 | `NemotronForCausalLM` | Nemotron-3, Nemotron-4, Minitron | `nvidia/Minitron-8B-Base`, `mgoin/Nemotron-4-340B-Base-hf-FP8`, etc. | ✅︎ | ✅︎ | ✅︎ |
-| `NemotronHForCausalLM` | Nemotron-H | `nvidia/Nemotron-H-8B-Base-8K`, `nvidia/Nemotron-H-47B-Base-8K`, `nvidia/Nemotron-H-56B-Base-8K`, etc. | ✅︎ | ✅︎ | |
+| `NemotronHForCausalLM` | Nemotron-H | `nvidia/Nemotron-H-8B-Base-8K`, `nvidia/Nemotron-H-47B-Base-8K`, `nvidia/Nemotron-H-56B-Base-8K`, etc. | ✅︎ | ✅︎ | ✅ |
```
```suggestion
| `NemotronHForCausalLM` | Nemotron-H | `nvidia/Nemotron-H-8B-Base-8K`, `nvidia/Nemotron-H-47B-Base-8K`, `nvidia/Nemotron-H-56B-Base-8K`, etc. | ✅︎ | ✅︎ | ✅︎ |
```

DarkLight1337 commented on docs/models/supported_models.md:
```diff
@@ -389,7 +389,7 @@ Specified using `--task generate`.
 | `MiniMaxM1ForCausalLM` | MiniMax-Text | `MiniMaxAI/MiniMax-M1-40k`, `MiniMaxAI/MiniMax-M1-80k`, etc. | | | |
 | `MiniMaxText01ForCausalLM` | MiniMax-Text | `MiniMaxAI/MiniMax-Text-01`, etc. | | | |
-| `Zamba2ForCausalLM` | Zamba2 | `Zyphra/Zamba2-7B-instruct`, `Zyphra/Zamba2-2.7B-instruct`, `Zyphra/Zamba2-1.2B-instruct`, etc. | | | |
+| `Zamba2ForCausalLM` | Zamba2 | `Zyphra/Zamba2-7B-instruct`, `Zyphra/Zamba2-2.7B-instruct`, `Zyphra/Zamba2-1.2B-instruct`, etc. | | | ✅ |
```
```suggestion
| `Zamba2ForCausalLM` | Zamba2 | `Zyphra/Zamba2-7B-instruct`, `Zyphra/Zamba2-2.7B-instruct`, `Zyphra/Zamba2-1.2B-instruct`, etc. | | | ✅︎ |
```

tdoublep committed:
Apply suggestions from code review

Unicode icon weirdness

Co-authored-by: Cyrus Leung <cyrus.tl.leung@gmail.com>
Signed-off-by: Thomas Parnell <tpa@zurich.ibm.com>
